web-marketplace/src/components/molecules/CreditsAmount/CreditsAmount.utils.tsx

```diff
@@ -1,37 +1,171 @@
+import { PaymentOptionsType } from 'pages/BuyCredits/BuyCredits.types';
+import { UISellOrderInfo } from 'pages/Projects/AllProjects/AllProjects.types';
+
+import { CURRENCY_AMOUNT } from './CreditsAmount.constants';
 
 export const getCreditsAvailablePerCurrency = (
-  currency: Currency,
-  creditDetails: CreditDetails[],
+  paymentOption: PaymentOptionsType,
+  filteredCryptoSellOrders: Array<UISellOrderInfo> | undefined,
+  cardSellOrders: Array<CardSellOrder>,
+  creditTypePrecision?: number | null,
 ) => {
-  return (
-    creditDetails.find(credit => credit.currency === currency)
-      ?.availableCredits || 0
-  );
+  return paymentOption === PAYMENT_OPTIONS.CARD
+    ? cardSellOrders.reduce((prev, cur) => prev + Number(cur.quantity), 0)
+    : filteredCryptoSellOrders?.reduce((prev, cur) => {
+        return parseFloat(
+          (prev + Number(cur.quantity)).toFixed(creditTypePrecision || 6),
+        );
+      }, 0) || 0;
 };
 
-export const getVintageCredits = (
-  creditVintageOptions: string[],
-  creditVintages: CreditsVintages[],
-) => {
-  return creditVintageOptions.reduce((sum: number, option: string) => {
-    const credits =
-      creditVintages.find(vintage => vintage.batchDenom === option)?.credits ||
-      '0';
-    return sum + +credits;
-  }, 0);
+export function getSpendingCap(
+  paymentOption: PaymentOptionsType,
+  filteredCryptoSellOrders: Array<UISellOrderInfo> | undefined,
+  cardSellOrders: Array<CardSellOrder>,
+) {
+  return paymentOption === PAYMENT_OPTIONS.CARD
+    ? cardSellOrders.reduce(
+        (prev, cur) => prev + Number(cur.quantity) * cur.usdPrice,
+        0,
+      )
+    : microToDenom(
+        filteredCryptoSellOrders?.reduce(
+          (prev, cur) => prev + Number(cur.quantity) * Number(cur.askAmount),
+          0,
+        ) || 0,
+      );
+}
+
+type GetCreditsAmountParams = {
+  value: number;
+  card: boolean;
+  orderedSellOrders: UISellOrderInfo[];
+  creditTypePrecision?: number | null;
+};
+export const getCreditsAmount = ({
+  value,
+  card,
+  orderedSellOrders,
+  creditTypePrecision,
+}: GetCreditsAmountParams) => {
+  const currentCurrencyAmount = card ? value : denomToMicro(value);
+  let currentCreditsAmount = 0;
+  let currencyAmountLeft = currentCurrencyAmount;
+  const sellOrders = [];
+
+  for (let i = 0; i < orderedSellOrders.length; i++) {
+    const order = orderedSellOrders[i];
+    const price = getSellOrderPrice({ order, card });
+    const quantity = Number(order.quantity);
+    const orderTotalAmount = quantity * price;
+
+    if (currencyAmountLeft >= orderTotalAmount) {
+      currencyAmountLeft = parseFloat(
+        (currencyAmountLeft - orderTotalAmount).toFixed(6),
+      );
+      currentCreditsAmount += quantity;
+      sellOrders.push(formatFullSellOrder({ order, card, price }));
+      if (currencyAmountLeft === 0) break;
+    } else {
+      currentCreditsAmount += currencyAmountLeft / price;
+      sellOrders.push({
```

Comment:
We could improve readability by extracting this into a function and potentially use it again in `getCurrencyAmount`: 
```
...
sellOrders.push(
  createOrderDetails(
    order,
    currencyAmountLeft,
    price,
    creditTypePrecision,
    card,
  ),
);
...


function createOrderDetails(
  order: UISellOrderInfo,
  currencyAmountLeft: number,
  price: number,
  creditTypePrecision: number | null | undefined,
  card: boolean,
): any {
  return {
    sellOrderId: order.id,
    quantity: (currencyAmountLeft / price).toFixed(creditTypePrecision || 6),
    bidPrice: !card
      ? { amount: String(price), denom: order.askDenom }
      : undefined,
    price: card ? price * 100 : undefined,
  };
}
```

Reply:
I've updated our existing `formatFullSellOrder` function to take an optional quantity as param
see https://github.com/regen-network/regen-web/pull/2477/commits/374be18423ad537000c3378df13032dec87460c9#diff-62c8723f280ca5d74f2f49f2027c2c5f403f8c8944c66d6b2eb960874b66c717